React better to partition migrations

mscastanho opened the pull request:
Partition migrations on PowerVM may take a few minutes, and during that
time libnxz will see consecutive paste failures, so we need to keep
waiting until the migration is complete. So use different timeouts for
waiting for a valid CSB and for a successful paste, giving PowerVM more
time on the paste case to properly handle migrations.

There's currently a bug in the kernel side in which the CSB might not
become valid for a long time during a migration unless we issue a new
paste. To work around it nx_submit_job has been changed to avoid a nasty
exit() call if nxu_run_job has returned with ETIMEDOUT. Instead, it now
just returns -1, triggering a job resubmission.

tuliom reviewed (commented):
I like what we have here, but I think we have to be careful with the return value from `nx_submit_job()`.

abalib commented on lib/nx_zlib.c:
```diff
@@ -488,8 +488,7 @@ int nx_submit_job(nx_dde_t *src, nx_dde_t *dst, nx_gzip_crb_cpb_t *cmdp, nx_devp
 			__FUNCTION__, __LINE__, ret);
-		/* Panic! Don't know what to do */
-		exit(-1);
+		return -1;
```
ETIME* are network return codes.  May be EAGAIN?  Your call better than mine
  https://man7.org/linux/man-pages/man3/errno.3.html

mscastanho replied:
The workbook mentions 240-249 and 251 error codes are `reserved for future system software use`. Would it be valid to create an internal CC error code for this case here?

abalib replied:
I think you will need to ask Haren and/or Bart.  It might be power architecture decision, may require RFC etc.  Haren stole one of the CC for the page faults.  He may know the procedure.

tuliom replied:
I do agree with `-EAGAIN`, but that isn't a valid CC value according to the NX GZIP user manual.
I'm starting to re-think my original proposal because `nx_submit_job()` returns an `int`, while CC is `uint8_t`.
In other words, it's already not a proper CC, as opposed to what I said before.
What if we document the return value as the following?
```
nx_submit_job() returns 0 in case of success. In case of hardware or OS-detected error, returns a positive value as reported in the Completion Code from the CSB.  Returns -EAGAIN in case of timeout.
```

abalib commented:
> Are you aware of any such situations?

I am not aware. I don't know of any situation where csb.v will not be set unless there is a bug or hardware error.  

Regarding "reissue" and assuming I understood your thinking:
I would be concerned about reissuing the job, if csb.v timed out.  What if the first job comes back after the second reissue?  Two jobs might stomp on each others buffers.  Does my question make sense here?

Is there a problem waiting for csb.v in an infinite loop?  Let's the sysadmin deal with stuck processes.

tuliom commented:
>  Does my question make sense here?

@abalib I think it does and it's only showing it's important to get a proper csb.v before taking more decisions.

> Is there a problem waiting for csb.v in an infinite loop?

That could cause starvation to other processes trying to use the NX GZIP while processes in an infinite loop lock the VAS credits.  IMHO, it's fair to add a high timeout value (i.e. minutes) and avoid this scenario.

mscastanho commented:
> That could cause starvation to other processes trying to use the NX GZIP while processes in an infinite loop lock the VAS credits. IMHO, it's fair to add a high timeout value (i.e. minutes) and avoid this scenario.

@tuliom But if we don't add the infinite loop, and if we can't re-issue the job for the reasons @abalib mentioned (which I think makes sense by the way), the only thing left to do is return an error to the user, like this PR is currently doing.

This could put applications that were written with zlib in mind in a difficult situation, because they wouldn't know how to proceed since this would be an unexpected error.

mscastanho commented:
@tuliom Fair enough. I updated the PR:
- Return -EAGAIN from nxu_run_job in case of timeout
- Added the documentation about the value returned by `nx_submit_job`
- Removed some dead code that would appear to return -EAGAIN to avoid confusion.
